src/data/loaders.py

```diff
@@ -1,113 +1,423 @@
-    """Turns a file into a list of "words"
+MAX_SHIFT_STEPS = 100
 
-    Arguments:
-        filepath (str): path to the lyrics file. e.g.
-            "/home/user/lyrics_data/tool/lateralus.txt"
-        max_len (int): the maximum length
+MIN_MIDI_VELOCITY = 1
+MAX_MIDI_VELOCITY = 127
+
+MIN_MIDI_PROGRAM = 0
+MAX_MIDI_PROGRAM = 127
+PROGRAMS_PER_FAMILY = 8
+
+
+log = logging.getLogger("few-shot")
+
+
+class Loader(object):
+    """A class for turning data into a sequence of tokens.
     """
-    token_count = 0
-    for line in open(filepath, 'r', errors='ignore'):
-        for token in line.split():
-            yield token
-            token_count += 1
-            if token_count >= max_len:
-                return
+    def __init__(self, max_len, dtype=np.int32, persist=True):
+        self.max_len = max_len
+        self.dtype = dtype
+        self.persist = persist
 
+    def is_song(self, filepath):
+        raise NotImplementedError
 
-def tokenize_midi_file(filepath):
-    """Turns a MIDI file into a list of event IDs.
+    def read(self, filepath):
+        raise NotImplementedError
 
-    Arguments:
-        filepath (str): path to the lyrics file. e.g.
-            "/home/user/freemidi_data/Tool/lateralus.mid"
+    def tokenize(self, data):
+        raise NotImplementedError
+
+    def detokenize(self, numpy_data):
+        raise NotImplementedError
+
+    def validate(self, filepath):
+        try:
+            self.load(filepath)
+            return True
+        except OSError:
+            return False
+        except KeyError:
+            return False
+        except EOFError:
+            return False
+        except IndexError:
+            return False
+        except ValueError:
+            return False
+
+    def load(self, filepath):
+        npfile = '%s.%s.npy' % (filepath, self.max_len)
+        if self.persist and os.path.isfile(npfile):
+            return np.load(npfile).astype(self.dtype)
+        else:
+            data = self.read(filepath)
+            tokens = self.tokenize(data)
+            numpy_tokens = np.zeros(self.max_len, dtype=self.dtype)
+            for token_index in range(min(self.max_len, len(tokens))):
+                numpy_tokens[token_index] = tokens[token_index]
+            if self.persist:
+                np.save(npfile, numpy_tokens)
+            return numpy_tokens
+
+
+class MIDILoader(Loader):
+    """Objects of this class parse MIDI files into a sequence of note IDs
     """
-    raise NotImplementedError
+    def read(self, filepath):
+        """Reads a MIDI file.
 
+        Arguments:
+            filepath (str): path to the lyrics file. e.g.
+                "/home/user/freemidi_data/Tool/lateralus.mid"
 
-class LyricsLoader(object):
-    """A class which loads and parses lyrics files into a list of word IDs.
+        """
+        return pretty_midi.PrettyMIDI(filepath)
+
+    def is_song(self, filepath):
+        return filepath.endswith('.mid')
+
+    def tokenize(self, midi):
```

Comment:
@sagelywizard @siddharth-agrawal is it possible to add documentation that states broadly how tokenization occurs for MIDI files in terms of the sequence of functions that are being called and some short documentation for what `get_notes`, `get_control_changes`, and `apply_sustain_control_changes` and `midi_control_changes` is doing?

Reply:
+1

Reply:
Sure.